File: python/239.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if len(nums) <= 1:
            return nums
        dqueue = []
        for i in range(k):
            while dqueue and dqueue[-1] < nums[i]:
                dqueue = dqueue[:-1]
            dqueue.append(nums[i])
        result = []
        result.append(dqueue[0])
        for j in range(k, len(nums)):
            if nums[j - k] == dqueue[0]:
                dqueue = dqueue[1:]
            while dqueue and dqueue[-1] < nums[j]:
                dqueue = dqueue[:-1]
            dqueue.append(nums[j])
            result.append(dqueue[0])
        return result
            
        

File: java/239.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length <= 1)
            return nums;
        Deque<Integer> dq = new LinkedList<>();
        for (int i = 0; i < k; i++) {
            while (!dq.isEmpty() && nums[i] > dq.peekLast())
                dq.removeLast();
            dq.add(nums[i]);
        }
        int[] result = new int[nums.length - k + 1];
        int index = 0;
         result[index++] = dq.peek();
        for (int j = k; j < nums.length; j++) {
            if (dq.peek() == nums[j - k])
                dq.removeFirst();
            while (!dq.isEmpty() && nums[j] > dq.peekLast())
                dq.removeLast();
            dq.add(nums[j]);
            result[index++] = dq.peek();
        }
        return result;
    }
}
